lightning/src/debug_sync.rs

```diff
@@ -0,0 +1,216 @@
+pub use ::alloc::sync::Arc;
+use core::ops::{Deref, DerefMut};
+use core::time::Duration;
+
+use std::collections::HashSet;
+use std::cell::RefCell;
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use std::sync::Mutex as StdMutex;
+use std::sync::MutexGuard as StdMutexGuard;
+use std::sync::RwLock as StdRwLock;
+use std::sync::RwLockReadGuard as StdRwLockReadGuard;
+use std::sync::RwLockWriteGuard as StdRwLockWriteGuard;
+use std::sync::Condvar as StdCondvar;
+
+#[cfg(feature = "backtrace")]
+use backtrace::Backtrace;
+
+pub type LockResult<Guard> = Result<Guard, ()>;
+
+pub struct Condvar {
+	inner: StdCondvar,
+}
+
+impl Condvar {
+	pub fn new() -> Condvar {
+		Condvar { inner: StdCondvar::new() }
+	}
+
+	pub fn wait<'a, T>(&'a self, guard: MutexGuard<'a, T>) -> LockResult<MutexGuard<'a, T>> {
+		let mutex: &'a Mutex<T> = guard.mutex;
+		self.inner.wait(guard.into_inner()).map(|lock| MutexGuard { mutex, lock }).map_err(|_| ())
+	}
+
+	#[allow(unused)]
+	pub fn wait_timeout<'a, T>(&'a self, guard: MutexGuard<'a, T>, dur: Duration) -> LockResult<(MutexGuard<'a, T>, ())> {
+		let mutex = guard.mutex;
+		self.inner.wait_timeout(guard.into_inner(), dur).map(|(lock, _)| (MutexGuard { mutex, lock }, ())).map_err(|_| ())
+	}
+
+	pub fn notify_all(&self) { self.inner.notify_all(); }
+}
+
+thread_local! {
+	/// We track the set of locks currently held by a reference to their `MutexMetadata`
+	static MUTEXES_HELD: RefCell<HashSet<Arc<MutexMetadata>>> = RefCell::new(HashSet::new());
+}
+static MUTEX_IDX: AtomicUsize = AtomicUsize::new(0);
+
+/// Metadata about a single mutex, by id, the set of things locked-before it, and the backtrace of
+/// when the Mutex itself was constructed.
+struct MutexMetadata {
+	mutex_idx: u64,
+	locked_before: StdMutex<HashSet<Arc<MutexMetadata>>>,
+	#[cfg(feature = "backtrace")]
+	mutex_construction_bt: Backtrace,
+}
+impl PartialEq for MutexMetadata {
+	fn eq(&self, o: &MutexMetadata) -> bool { self.mutex_idx == o.mutex_idx }
+}
+impl Eq for MutexMetadata {}
+impl std::hash::Hash for MutexMetadata {
+	fn hash<H: std::hash::Hasher>(&self, hasher: &mut H) { hasher.write_u64(self.mutex_idx); }
+}
+
+pub struct Mutex<T: Sized> {
+	inner: StdMutex<T>,
+	deps: Arc<MutexMetadata>,
+	skip_checks: bool
+}
+
+#[must_use = "if unused the Mutex will immediately unlock"]
+pub struct MutexGuard<'a, T: Sized + 'a> {
+	mutex: &'a Mutex<T>,
+	lock: StdMutexGuard<'a, T>,
+}
+
+impl<'a, T: Sized> MutexGuard<'a, T> {
```

Comment:
Should these be `?Sized`?

Reply:
I tried that originally but the compiler at some point got mad so I just dropped it. Works for all our current usages and isn't pub so didn't bother looking into it.

Reply:
Gotcha. Wasn't sure if is was related to the unsafe code.